Extract interfaces for vms, hosts providers

imtayadeway opened the pull request:
Extracts interfaces for vms, hosts and providers. Concrete types have been provided for vmware for illustration - these can later be separated/extracted out.

Queries can be performed on the concrete types (e.g. `query { vmware_vms { edges { node { name }}}}`) or across all implementations of the interface (`query { vms { edges { node { name }}}}`). Non-substitutable specializations in concrete types can be retrieved in the latter query by use of fragments (see http://graphql.org/learn/queries/#inline-fragments).

TODO: try to figure out how the `resolve_type` function on the schema can not have to know about provider-specific concrete types. May leave this as a follow up.

Resolves https://www.pivotaltracker.com/story/show/157486444

imtayadeway committed:
Extract concrete type VmVmware and make Vm an interface

imtayadeway committed:
Extract concrete type HostVmware and make Host an interface

imtayadeway committed:
Extract concrete type ProviderVmware and make Provider an interface

chrisarcand commented:
Nice, going down the right path.

One thing I see right away is you have `vmware_hosts` as a field but it's `HostVmware` as the type. Pick one or the other! 

* I like VmwareHost because it reads better but....
* ...see concerns over naming things; Shopify, for example, does things backwords because it lexically keeps things together. (instead of `PlaceOrder` and `CancelOrder` as 'action sounding things', they do `OrderPlace` and `OrderCancel` so the files stay close together)

imtayadeway committed:
Add some consistency to naming

"vmware_vms" *sounds* more natural but makes things like
sorting/grouping harder. Ultimately I think it would be better to
optimize for the latter.

Fryguy commented:
Lexical ordering is exactly why models are named the way they are.
